### PR TITLE
Refactor `TestSceneMatchStartControl` to avoid usage of `TestMultiplayerClient`

### PR DESCRIPTION
**Pull request by peppy**
A few notes:

- The goal here is to test the `MatchStartComponent` rather than the behaviour of `MultiplayerClient` itself. I've added the bare minimum implementation intentionally (for instance, changing setting to autostart will not actually start a countdown. this could be added if we also want to be able to visually play with the control better, open to feedback).
- The mocked object is currently `MultiplayerClient`. That class still has some local logic, but none is relied upon by this test scene. This allows switching to an interface or potentially combining the implementation into a single class (if we ever reach a point where `TestMultiplayerClient` is not required).
- I've finished this without revisiting `CurrentPlaylistItem` woes (aka https://github.com/ppy/osu/pull/17702). Some of the room-specific logic can likely be removed when refactoring is done around that, as discussed today in voice.
- An argument could be made that this isn't testing the fact that things don't happen immediately with `MultiplayerClient` (scheduled + server delays). If this is something we want to account for, it's quite easy to add `Schedule` or `Task.Delay`s to the moq implementations. That said, I'm not sure that this is something we should be testing for on a per-component level.

Wanted to attempt this one to see how easily the moq approach will work with a more complicated component.

I'm pretty happy with how this turned out, and think it's a good direction for testing individual components going forward 🤞

### CODE REVIEW

File: osu.Game/Tests/Visual/Multiplayer/TestMultiplayerClient.cs [hunk 1]
```diff
@@ -360,62 +329,6 @@ public override async Task SendMatchRequest(MatchUserRequest request)
             }
         }
 
-        private void startCountdown(MultiplayerCountdown countdown, Func<Task> continuation)
```
Comment (smoogipoo) [inline osu.Game/Tests/Visual/Multiplayer/TestMultiplayerClient.cs:363]: Removing this means that countdowns are no longer testable in "flow" tests (`TestSceneMultiplayer`).

File: osu.Game.Tests/Visual/Multiplayer/TestSceneMatchStartControl.cs [hunk 2]
```diff
@@ -3,80 +3,155 @@
...
+                                 raiseRoomUpdated();
+                             });
+
+            multiplayerClient.Setup(m => m.StartMatch())
+                             .Callback(() =>
+                             {
+                                 multiplayerClient.Raise(m => m.LoadRequested -= null);
+
+                                 // immediately "end" gameplay, as we don't care about that part of the process.
+                                 changeUserState(localUser.UserID, MultiplayerUserState.Idle);
```
Comment (smoogipoo) [inline osu.Game.Tests/Visual/Multiplayer/TestSceneMatchStartControl.cs:79]: We do care about that part of the process. It's skipping an entire sample playback plus testing for the disabled state of the button during gameplay. Admittedly, there's no test for it (was probably removed somewhere with the `ClickButtonWhenEnabled` changes, but it needs to happen.

Comment (smoogipoo) [reply osu.Game.Tests/Visual/Multiplayer/TestSceneMatchStartControl.cs:79]: Furthermore, this should be stopping the countdown otherwise it gets stuck in this state:
image

**Review by smoogipoo — CHANGES_REQUESTED** [review]

File: osu.Game.Tests/Visual/Multiplayer/TestSceneMatchStartControl.cs (hunk 2 above)
Comment (peppy) [reply osu.Game.Tests/Visual/Multiplayer/TestSceneMatchStartControl.cs:79]: Which sample playback do you mean?

From the stuck state, you should be able to cancel or start a new countdown – is that not working?

File: osu.Game/Tests/Visual/Multiplayer/TestMultiplayerClient.cs (hunk 1 above)
Comment (peppy) [reply osu.Game/Tests/Visual/Multiplayer/TestMultiplayerClient.cs:363]: Is the goal here to test it ad-hoc interactively? Would you be against making an "InteractiveMultiplayer" test scene which uses the online (dev) server for said purposes?

File: osu.Game.Tests/Visual/Multiplayer/TestSceneMatchStartControl.cs (hunk 2 above)
Comment (smoogipoo) [reply osu.Game.Tests/Visual/Multiplayer/TestSceneMatchStartControl.cs:79]: > Which sample playback do you mean?

Hmm, I was referring to the gameplay start sound, but that one's actually being played by the subscreen.

> you should be able to cancel or start a new countdown

Yes, this works.

File: osu.Game/Tests/Visual/Multiplayer/TestMultiplayerClient.cs (hunk 1 above)
Comment (smoogipoo) [reply osu.Game/Tests/Visual/Multiplayer/TestMultiplayerClient.cs:363]: Let's not use online anything, please. I am still traumatised by the CI test failures when online stuff was used (simple things like timeouts to httpbin or a.ppy.sh), and is the reason why this has been reimplemented client-side.

The test is testing actual functionality.

Comment (peppy) [reply osu.Game/Tests/Visual/Multiplayer/TestMultiplayerClient.cs:363]: For sure, I was proposing *only* for use in interactive visual tests (ie during development). Not replacing usage in this test.

Comment (peppy) [reply osu.Game/Tests/Visual/Multiplayer/TestMultiplayerClient.cs:363]: That said, I'm fine with leaving said code in place until I have a proposal which fully replaces it, if that's preferred. I removed it mostly to gauge response (given that it's not used anywhere else and one of the most complex pieces of `TestMultiplayerClient`.

Comment (smoogipoo) [reply osu.Game/Tests/Visual/Multiplayer/TestMultiplayerClient.cs:363]: So your thinking is as long as it isn't used in automated testing, it isn't required? I guess I could get along with that.

Personally I don't find this test scene all that useful interactively, opting to instead test the full client + server flow via in-game.

Comment (peppy) [reply osu.Game/Tests/Visual/Multiplayer/TestMultiplayerClient.cs:363]: My current line of thought is that:

- We need to test at component boundaries more often. Every expectation of the component should be tested in isolation (sometimes we do that, other times we don't, this is mostly a proposal for going forward).
- For "flow" tests, where more than one components are involved in cross-component interactions, they should be tested together in an isolated scene, but still not require the multiplayer client/server.
- We need test coverage of `MultiplayerClient` on its own, with the input and output expectations of it. This will hopefully reduce the perceived need for some of the tests in `TestSceneMultiplayer`.

I'm planning to continue taking this a step at a time (working through every scene that is using `OnlinePlayTestScene`) to reduce the cross-dependencies of tests, as that seems to be the biggest point of stress in the system.

As you say, for development it's easy to spin up a new tests scene and get the online support when required to test the "overall picture", but I think we need to do that less in tests and try and isolate at boundaries more, splitting up responsibilities of individual pieces to reduce the overall complexity (and make refactoring the pieces and related tests easier).

Still formulating things and doing some reading on good testing practices, but that's where I am as of today.

**Review by smoogipoo — APPROVED** [review]

**Comment by smoogipoo** [comment]
@bdach want your approval on the structure here as well.

**Comment by bdach** [comment]
Am generally fine with this, albeit there are a few places where the mocks reimplement parts of server to a degree (see the countdown handling. as well as handling unreadying users on room settings change). Not sure how that's going to pan out in the long run, but also willing to give this a go in the here-and-now and see what end result is arrived at.

**Review by bdach — APPROVED** [review]

**Comment by peppy** [comment]
> Am generally fine with this, albeit there are a few places where the mocks reimplement parts of server to a degree (see the countdown handling. as well as handling unreadying users on room settings change). Not sure how that's going to pan out in the long run, but also willing to give this a go in the here-and-now and see what end result is arrived at.

The unreadying was only to make one existing test work. I was considering removing said test coverage but ended up deciding to match 1:1 to show how it *could* be done.
